### PR TITLE
[#2] Implement safe list

### LINKED ISSUE #2: Implement Safelist contract
# Clarification and motivation
The specification defines [`Safelist`](https://github.com/serokell/tezos-globacap/blob/master/docs/specification.md#safelist-contract) contract which is going to be used for regulatory purposes.

As an additional requirement, we want to use Indigo.
# Acceptance criteria
`Safelist` contract is implemented and tested.

### PR DESCRIPTION
**Pull request by rvem**
## Description

Implementation for the first part of specification: `Safelist` contract.
Along with tests and documentation.

## Related issue(s)

Resolves #2

#### Related changes (conditional)

- Tests
  - [x] If I added new functionality, I added tests covering it.
  - [x] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

- Documentation
  - [x] I checked whether I should update the docs and did so if necessary:
    - [README](../tree/master/README.md)
    - Haddock

#### Stylistic guide (mandatory)

- [x] My commits comply with [the policy used in Serokell](https://www.notion.so/serokell/Where-and-how-to-commit-your-work-58f8973a4b3142c8abbd2e6fd5b3a08e).
- [x] My code complies with the [style guide](../tree/master/docs/code-style.md).

### CODE REVIEW

**Comment by gromakovsky** [comment]
> My code complies with the style guide.

This link is broken.

File: test/Test/Indigo/Contracts/Safelist.hs [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+    integrationalTestExpectation $ do
+      sl <- originateSafelist ownerAddress
+      withSender ownerAddress $ lCallDef sl $ AddToWhitelist genesisAddress1
+      lCallDef sl $ EnsureSafelistConstraints
+        ( #isWhitelisted .! [genesisAddress1, genesisAddress2]
+        , #isNonBlacklisted .! [genesisAddress1, genesisAddress2]
+        )
+      validate . Left $
+        lExpectCustomError #constraintUnsatisfied (genesisAddress2, [mt|nonWhitelisted|])
+  , testCase "Fails on first unsatisfied constraint ver 1" $
```
[the remark this answers is not in the document]
Comment (rvem) [reply test/Test/Indigo/Contracts/Safelist.hs]: Mda, copypaste is evil

File: src/Indigo/Contracts/Safelist.hs [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+        ensureSenderIsAdmin
+        addOrRemoveFromStorageSet addr #whitelist Remove
+    , #cIsWhitelisted /-> \v -> presentedInStorageSet v #whitelist
+    , #cAddToBlacklist /-> \addr -> do
+        ensureSenderIsAdmin
+        addOrRemoveFromStorageSet addr #blacklist Insert
+    , #cRemoveFromBlacklist /-> \addr -> do
+        ensureSenderIsAdmin
+        addOrRemoveFromStorageSet addr #blacklist Remove
+    , #cIsBlacklisted /-> \v -> presentedInStorageSet v #blacklist
```
[the remark this answers is not in the document]
Comment (rvem) [reply src/Indigo/Contracts/Safelist.hs]: :thonk:

[the remark this answers is not in the document]
Comment (rvem) [reply src/Indigo/Contracts/Safelist.hs]: One problem, that now there is `doc` section here

File: src/Indigo/Contracts/Safelist.hs [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+    , #cEnsureSafelistConstraints /-> \constraints ->
+        ensureSafelistConstraints constraints
+    )
+
+storage :: HasStorage Storage => Var Storage
+storage = storageVar
```
[the remark this answers is not in the document]
Comment (rvem) [reply src/Indigo/Contracts/Safelist.hs]: The problem is that `storageVar` cannot automatically infer `Storage` type and you have to right
`storageVar @Storage` everywhere in the code

[the remark this answers is not in the document]
Comment (rvem) [reply src/Indigo/Contracts/Safelist.hs]: https://gitlab.com/morley-framework/morley/-/merge_requests/297#note_297654734

File: src/Indigo/Contracts/Safelist.hs (hunk 2 above)
[the remark this answers is not in the document]
Comment (gromakovsky) [reply src/Indigo/Contracts/Safelist.hs]: Ah, ok, fine then, I was reviewing it on per-commit basis and didn't realize that.

File: src/Indigo/Contracts/Safelist.hs (hunk 3 above)
[the remark this answers is not in the document]
Comment (gromakovsky) [reply src/Indigo/Contracts/Safelist.hs]: OK

**Pull request merged**